### PR TITLE
add nightly build and push

### PR DESCRIPTION
**Pull request by tuunit**
## Description
Add nightly docker image build and publish workflow for the oauth2-proxy master branch. To always have a "bleeding edge" build of the latest merges for the community to test.

This PR requires the addition of the repository secrets for a service account with access to the quay.io repo:

- `REGISTRY_USERNAME_NIGHTLY`
- `REGISTYR_PASSWORD_NIGHTLY`

## Motivation and Context
As the release cycle is rather slow some features might only be available on the master branch for weeks or month at a time. Therefore it would be great to have nightly builds of the master branch available.

## How Has This Been Tested?
Build and publish workflow has been executed locally and on a forked repo.

## Further notes
I took the liberty to update the doc images as part of this PR.

## Checklist:

- [X] My change requires a change to the documentation or CHANGELOG.
- [X] I have updated the documentation/CHANGELOG accordingly.
- [X] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

**Review by kvanzuijlen — COMMENTED** [review]
quay.io registry needs to be updated, other than that I think its LGTM

**Comment by tuunit** [comment]
> quay.io registry needs to be updated, other than that I think its LGTM

@JoelSpeed let me know when you have created the new registry repo and I'll update it accordingly.

File: .github/workflows/nightly.yml [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        username: ${{ secrets.REGISTRY_USERNAME }}
+        password: ${{ secrets.REGISTRY_PASSWORD }}
+
+    - name: Build images
+      run: |
+        make docker-nightly-build
+
+    - name: Push images
+      run: |
+        make docker-nightly-push
```
Comment (JoelSpeed) [inline .github/workflows/nightly.yml:39]: Missing new line at the end of this file

File: Makefile [hunk 2]
```diff
@@ -4,7 +4,8 @@ GOLANGCILINT ?= golangci-lint
 BINARY := oauth2-proxy
 VERSION ?= $(shell git describe --always --dirty --tags 2>/dev/null || echo "undefined")
 # Allow to override image registry.
-REGISTRY ?= quay.io/oauth2-proxy
+REGISTRY ?= quay.io/tuunit
```
Comment (JoelSpeed) [inline Makefile:7]: Needs fixing

**Comment by JoelSpeed** [comment]
Are we creating a new repository or just using a nightly tag on the current repository? What do we want to create on Quay?

**Comment by tuunit** [comment]
> Are we creating a new repository or just using a nightly tag on the current repository? What do we want to create on Quay?

I think I would prefer to push the nightly builds to a separate repository. Otherwise it will just clutter the main repository.

**Comment by JoelSpeed** [comment]
Ok, I'll set up a repo `quay.io/oauth2-proxy/oauth2-proxy-nightly` and we can use that one for the nightlies, will then set up the env vars as you've suggested

**Comment by tuunit** [comment]
@JoelSpeed I updated the branch and the Makefile to point to the correct repository again.

File: .github/workflows/nightly.yml [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+      uses: docker/setup-qemu-action@v2
+
+    - name: Set up Docker Buildx
+      id: buildx
+      uses: docker/setup-buildx-action@v2
+
+    - name: Login to quay.io
+      uses: docker/login-action@v2
+      with:
+        registry: quay.io/tuunit
```
Comment (JoelSpeed) [inline .github/workflows/nightly.yml:29]: This looks like it needs updating

Comment (tuunit) [reply .github/workflows/nightly.yml:29]: Yep, obviously I missed that...

**Review by JoelSpeed — APPROVED** [review]

**Review by JoelSpeed — APPROVED** [review]

**Comment by ianroberts** [comment]
> Build and publish workflow has been executed locally and on a forked repo.

Now this is merged to master it might be a good idea to disable the nightly workflow from running in forks, since (a) I doubt you want forked repos trying to push to your `quay.io` repo and (b) the workflow runs will probably all fail anyway as the forks won't have access to the credential secrets.

I've suggested the obvious fix in #2330

**Pull request merged**